torch/csrc/jit/interpreter_autograd_function.h

```diff
@@ -1,36 +1,107 @@
+
 namespace torch { namespace jit {
+
+struct StageDetails {
+  std::vector<tracer::VariableFlags> input_flags;
+  std::vector<tracer::VariableFlags> output_flags;
+  std::vector<int> copied_next_fns;
+};
+
 struct InterpreterAutogradFunction : public autograd::Function {
-  InterpreterAutogradFunction(const jit::Code & code)
-  : interp_(code) {}
-  InterpreterAutogradFunction(const InterpreterState & interp_, autograd::FunctionFlags && f)
-  : autograd::Function(std::move(f)), interp_(interp_) {}
+  InterpreterAutogradFunction(const jit::Code & code,
+                              const std::vector<StageDetails>& stage_details)
+    : interp_(code)
+    , stage_details_(stage_details)
+    , stage_(0) {}
+
+  InterpreterAutogradFunction(InterpreterState interp,
+                              const std::vector<StageDetails>& stage_details,
+                              std::size_t stage)
+    : interp_(std::move(interp))
+    , stage_details_(stage_details)
+    , stage_(stage) {}
 
   virtual void willReleaseVariables() override {
-    keep_graph = false;
+    keep_graph_ = false;
   }
+
   virtual autograd::variable_list apply(const autograd::variable_list& inputs) override {
-    std::vector<at::Tensor> tinputs;
-    std::vector<at::Tensor> toutputs;
-    for(auto & i : inputs) {
-      tinputs.push_back(i.data());
+    // Initial correctness checks.
+    if (stage_ == stage_details_.size()) {
+      throw std::runtime_error(std::string("Function compiled only for ") +
+          std::to_string(stage_details_.size() - 1) + " derivatives. Use nderivs argument " +
+          "to request more.");
     }
-    InterpreterState interp = (keep_graph) ? interp_.clone() : interp_;
-    keep_graph = true;
+    if (used_) throw std::runtime_error(autograd::ERR_BACKWARD_TWICE);
+    used_ |= !keep_graph_;
+
+    const auto & details = stage_details_[stage_];
+
+    // Validate inputs
+    for (std::size_t i = 0; i < (std::size_t)num_inputs; ++i) {
+      if (!details.input_flags[i].verify(inputs[i])) {
+        throw std::runtime_error("JIT interpreter received inputs with different "
+            "flags than it was compiled for.");
+      }
+    }
+
+    // Run the interpreter
+    auto tinputs = fmap(inputs, [](const autograd::Variable& i) { return i.data(); });
+    std::vector<at::Tensor> toutputs;
+    InterpreterState interp = (keep_graph_) ? interp_.clone() : interp_;
     interp.runOneStage(tinputs, toutputs);
-    auto r = autograd::wrap_outputs(inputs, std::move(toutputs), [&](autograd::FunctionFlags f) {
-      return std::make_shared<InterpreterAutogradFunction>(interp, std::move(f));
-    });
-    return r;
+
+    // Lazily create grad_fn
+    std::shared_ptr<Function> grad_fn;
+    auto make_grad_fn = [&]() {
+      grad_fn = std::make_shared<InterpreterAutogradFunction>(
+          std::move(interp), stage_details_, stage_ + 1);
+      // Patch next_functions to include prevous stage next_functions
```

Comment:
Can you explain this next_functions passthrough patching to me?

Reply:
Sure. The problem is that e.g. stage 2 has to emit gradients not only for inputs to stage 1, but for inputs to stage 0 as well - those need to be copied to `next_functions` of stage 2 function from `next_functions` of stage 1 function.

Reply:
You can see that e.g. in expect file of [`test_backward`](https://github.com/pytorch/pytorch/blob/master/test/expect/TestJit.test_backward.expect) - stage 2 has 3 outputs.

Reply:
Let's add a comment here explaining what the expected inputs/outputs are for each stage. It is very subtle where the extra outputs come from.